feat(perpetual-proposer): Initialize PerpetualFundingRateProposer package

nicholaspai opened the pull request:
Signed-off-by: Nick Pai <npai.nyc@gmail.com>

<!--
  Thanks for submitting a pull request!
  We appreciate you spending the time to work on these changes. Please provide enough information so that others can review your pull request. The three fields below are mandatory.
-->

<!--
  Title
  Please include a concise title that briefly describes the change.
  Titles should follow https://www.conventionalcommits.org/.
  They should also be in the present simple tense.

  Examples:

  feat(dvm): adds a new function to compute voting rewards offchain
  fix(monitor): fixes broken link in liquidation log
  feat(voter-dapp): adds countdown timer component to the header
  build(solc): updates solc version to 0.6.12
  improve(emp-client): parallelizes web3 calls to improve performance

  For examples of other types (feat, fix, build, improve) and what they mean, take a look at the angular list:
  https://github.com/angular/angular/blob/22b96b9/CONTRIBUTING.md#type

  See https://github.com/UMAprotocol/protocol/blob/master/CONTRIBUTING.md#conventional-commits for more details on PR
  title expectations.
-->


**Motivation**

#2622
#2623
#2624
#2642

**Summary**

- Adds new event client that tracks `createdExpiringMultiParty` and `createdPerpetual` events emitted by financial contract factories
- Perpetual proposer bot can use this client to get a list of all live perpetuals
- Proposer bot has update function to fetch on-chain state and store it in its state cache, and `updateFundingRates()` method which proposes new rates where appropriate.

Additionally, refactors were made for logic that is shared between bot packages:
- `setAllowance` is now in `financial-templates-lib/helpers/allowance.js`. Used by all bots.
- `computeDeviationError` is now in `financial-templates-lib/helpers/math.js`. Used by SyntheticPegMonitor and RangeTrader.
- `isDeviationOutsideErrorMargin` is now in `financial-templates-lib/helpers/math.js` Used by OptimisticOracleProposer and FundingRateProposer.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ ]  Ran end-to-end test, running the code as in production
- [X]  New unit tests created
- [ ]  Existing tests adequate, no new tests required
- [ ]  All existing tests pass
- [ ]  Untested


**Issue(s)**

<!-- This PR must fix or refer to one or more issues. Please list them here. -->
Fixes #2622 
Fixes #2623
Fixes #2624
Fixes #2642

chrismaree commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,109 @@
+    };
+
+    // Look for events on chain from the previous seen block number to the current block number.
+    const eventToSearchFor = this.contractType === "Perpetual" ? "CreatedPerpetual" : "CreatedExpiringMultiParty";
```
nice

chrismaree commented on packages/financial-templates-lib/test/truffle/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,248 @@
+// Note: This is placed within the /truffle folder because EmpCreator/PerpCreator.new() fails due to incorrect library linking by hardhat.
```
rip.

chrismaree commented on packages/perpetual/index.js:
```diff
@@ -0,0 +1,179 @@
+#!/usr/bin/env node
```
2 initial nits:
1) do you want to call the package `Funding-Rate-Proposer` or maybe `Perpetual-Proposer`? IDK, Perpetual does not feel like a good name.
2) do you want to turn this into typescript, seeing this is a new package?

nicholaspai replied:
1. Renamed to `funding-rate-proposer`
2. Yes, but I wanted to get this out as quickly as possible. I'm thinking of either converting to TS in a subsequent PR or just biting bullet and doing it now. Do you think TS and truffle work well together?

mrice32 replied:
On the typescript question, I think I would be in favor of starting with ts if we hadn't already written most of the code. However, since it's already written in js, I would say we should just _convert_ to ts subsequently in the name of unblocking the perp rollout.

chrismaree commented on packages/perpetual/index.js:
```diff
@@ -0,0 +1,179 @@
+    // Else, if running in loop mode (pollingDelay != 0), then it should send a `info` level log.
+    logger[pollingDelay === 0 ? "debug" : "info"]({
+      at: "PerpetualFundingRateProposer#index",
+      message: "Perpetual funding rate proposer started 🌊",
```
new emoji please.

chrismaree commented on packages/perpetual/src/fundingRateProposer.js:
```diff
@@ -0,0 +1,223 @@
+  /**
+   * @notice Constructs new Perpetual FundingRate Proposer bot.
+   * @param {Object} logger Module used to send logs.
+   * @param {Object} perpetualFactoryClient Module used to query for live perpetual contracts
```
```suggestion
   * @param {Object} perpetualFactoryClient Module used to query for live perpetual contracts.
```

chrismaree commented on packages/perpetual/src/fundingRateProposer.js:
```diff
@@ -0,0 +1,223 @@
+  async update() {
+    await Promise.all([this.perpetualFactoryClient.update(), this.gasEstimator.update()]);
+
+    // Create contract object for each perpetual address created.
```
nice.

chrismaree commented on packages/perpetual/src/fundingRateProposer.js:
```diff
@@ -0,0 +1,223 @@
+  async updateFundingRates() {
+    this.logger.debug({
+      at: "PerpetualProposer#updateFundingRates",
+      message: "Checking for contract funding rates to update"
```
do you want to add:
```
perpetualsChecked: Object.keys(this.contractCache)
```
to the log to the addresses? might help a bit with debugging.

chrismaree commented on packages/perpetual/src/fundingRateProposer.js:
```diff
@@ -0,0 +1,223 @@
+    Object.assign(this, configWithDefaults);
+  }
+
+  async update() {
```
do you not need to update the price feeds within `this.priceFeedCache`?

chrismaree replied:
I see you do this in line 129 per feed. would we not want to do them all at once in this update method?

nicholaspai replied:
This is tough. I think the code would be cleaner if the `update()` for all pricefeeds was called in this `update()` method, but I left them for the `_updateFUndingRate()` method so that the pricefeeds could be updated as close as possible to their usage, to minimize gaps between when prices are fetched and whent hey are compared against on chain prices.

chrismaree commented on packages/perpetual/src/fundingRateProposer.js:
```diff
@@ -0,0 +1,223 @@
+    }
+
+    // With pricefeed successfully constructed, get the current funding rate
+    await priceFeed.update();
```
do we not want to group the updates into a `Promise.all` in the `update` method?

nicholaspai replied:
see [this comment](https://github.com/UMAprotocol/protocol/pull/2632#discussion_r586364110)

chrismaree commented on packages/perpetual/src/fundingRateProposer.js:
```diff
@@ -0,0 +1,223 @@
+
+    // Create a new pricefeed for this identifier. We might consider caching these price requests
+    // for re-use if any requests use the same identifier.
+    let newPriceFeed = await createReferencePriceFeedForFinancialContract(
```
this is super fresh. good job.

chrismaree reviewed (commented):
This is pretty awesome! great work on getting these done so quickly. I have a few nits on general updating methods and a few others. I will do a second more detailed review once these ones have been cleared.

nicholaspai commented:
@UMAprotocol/eng I've updated this PR to be less of an MVP and more of a fully fleshed-out new bot.

The main frustration I've been having is that using the `emp/perpCreators` in hardhat has given me trouble because of linked library issue. I'm not sure if there is an easy fix to make these tests run in hardhat, but perhaps that should be left for the next PR.

Here's a comment I've left in the tests:

```
// Note: This is placed within the /truffle folder because EmpCreator/PerpCreator.new() fails due to incorrect library linking by hardhat.
// `Error: ExpiringMultiPartyCreator contains unresolved libraries. You must deploy and link the following libraries before
//         you can deploy a new version of ExpiringMultiPartyCreator: $585a446ef18259666e65e81865270bd4dc$`
// We should look more into library linking via hardhat within a script: https://hardhat.org/plugins/hardhat-deploy.html#handling-contract-using-libraries
```

nicholaspai commented:
> @UMAprotocol/eng I've updated this PR to be less of an MVP and more of a fully fleshed-out new bot.
> 
> The main frustration I've been having is that using the `emp/perpCreators` in hardhat has given me trouble because of linked library issue. I'm not sure if there is an easy fix to make these tests run in hardhat, but perhaps that should be left for the next PR.
> 
> Here's a comment I've left in the tests:
> 
> ```
> // Note: This is placed within the /truffle folder because EmpCreator/PerpCreator.new() fails due to incorrect library linking by hardhat.
> // `Error: ExpiringMultiPartyCreator contains unresolved libraries. You must deploy and link the following libraries before
> //         you can deploy a new version of ExpiringMultiPartyCreator: $585a446ef18259666e65e81865270bd4dc$`
> // We should look more into library linking via hardhat within a script: https://hardhat.org/plugins/hardhat-deploy.html#handling-contract-using-libraries
> ```

Turns out I was able to move all tests to hardhat pretty easily by importing the `truffle-fixture.js` file from `@uma/core`

mrice32 commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,109 @@
+    financialContractFactoryAddress,
+    startingBlockNumber = 0,
+    endingBlockNumber = null,
+    contractType = "Perpetual"
```
One quick note: we could make this the name of the factory contract (ExpiringMultiPartyCreator or PerpetualCreator). Then that could be pulled directly from the artifact. Not strongly opinionated, but that seems a bit more obvious since we call these contracts different names (EMP vs ExpiringMultiParty vs ExpiribgMultiPartyCreator).

nicholaspai replied:
Definitely

mrice32 commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,109 @@
+    this.lastBlockToSearchUntil = endingBlockNumber;
+    this.lastUpdateTimestamp = 0;
+
+    this.contractType = contractType;
```
Should we check that the input is a valid type? What happens if someone passes "Perpteual"?

nicholaspai replied:
Done good catch

mrice32 commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,109 @@
+    this.createdContractEvents = [];
+  }
+
+  getAllCreatedContractEvents() {
```
Do you think it might be nice to replace this (or have a separate method) that returns just a list of deployed addresses rather than the full event?

nicholaspai replied:
Yes, but I don't really see a difference for now. If we find that we need this more (perhaps for the monitor) its easy enough to add the method. For now its just simpler to not do any filtering on the raw event data

nicholaspai replied:
Perhaps we should rename this client to just a `FinancialContractFactoryClient` and we can use it to precompute some other contract data, such as `fundingRateIdentifier`, `collateralCurrency`, etc., most of which we do in the `funding-rate-proposer/proposer/src` code right now. WDYT?

mrice32 replied:
I see. So this is modeled after the other event clients. I could imagine it pulling down other contract metadata, but I could also imagine that being overkill for some/most of the other use cases.

The reason I think it could be nice to return a list of contract addresses is that that seems pretty universal. Put another way, if we wanted a liquidator bot to monitor _all_ emps, it would use this client for the same purpose: grabbing all addresses. Some other metadata may be useful in some cases, but certainly the fundingRateIdentifier wouldn't be needed by the liquidator bot, for example, and it might be a waste to query for that.

Still not strongly opinionated -- just a thought about generalizability.

mrice32 replied:
Nothing necessarily needs to be changed here. Just want to make sure you see this thread.

nicholaspai replied:
> I see. So this is modeled after the other event clients. I could imagine it pulling down other contract metadata, but I could also imagine that being overkill for some/most of the other use cases.

Yeah problem would be if a user just wants contract addresses and no state, precomputing the state would slow the address query down.

I ultimately agree with you that this methid would be useful to return all addresses, so I'll rename this client to signify that its not just an event client, and add the method.

nicholaspai replied:
Done.

mrice32 commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,109 @@
+  }
+
+  async update() {
+    // The last block to search is either the value specified in the constructor (useful in serverless mode) or is the
```
Will we use this this way? Does this mean we'll only get contracts that are deployed within the last block window? Or are we effectively going to be searching from block 0 in serverless mode?

nicholaspai replied:
This is a great point, I'm thinking the generalized solution is to make the user of this client (i.e. the FundingRateProposer), just set `startingBlock = 0` when constructing this client.

mrice32 replied:
Yeah, I actually think this could have a cool double-use. In the monitor, we may want to know about _new_ EMPs. Here, we want to know about _all_ EMPs.

mrice32 commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,360 @@
+      this._setAllowances(),
+      // For each contract, create, save, and update a pricefeed instance for its identifier,
+      // or just update the pricefeed if the identifier already has cached an instance.
+      await this._cacheAndUpdatePriceFeeds()
```
```suggestion
      this._cacheAndUpdatePriceFeeds()
```

nicholaspai replied:
This is an async method because `createRefernecePriceFeed...` is async

mrice32 replied:
But it's being called inside a `Promise.all`, no? _setAllowances above is also async, but you don't put an await in front of it because the promise.all handles that if I'm reading this correctly.

nicholaspai replied:
oh true, done

mrice32 replied:
Bump.

nicholaspai replied:
This is done.

mrice32 commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,360 @@
+  // Sets allowances for all collateral currencies used live perpetual contracts.
+  async _setAllowances() {
+    // Get new approval receipts or null if approval was unneccessary.
+    const newApprovals = await Promise.map(Object.keys(this.contractCache), contractAddress => {
+      // The Perpetual requires approval to transfer the contract's collateral currency in order to post a bond.
+      // We'll set this once to the max value and top up whenever the bot's allowance drops below MAX_INT / 2.
+      return setAllowance(
+        this.web3,
+        this.gasEstimator,
+        this.account,
+        contractAddress,
+        this.contractCache[contractAddress].collateralAddress
+      );
+    });
+    newApprovals.forEach(receipt => {
+      if (receipt) {
+        this.logger.info({
+          at: "PerpetualProposer",
+          message: "Approved Perpetual contract to transfer unlimited collateral tokens 💰",
+          perpetual: receipt.spenderAddress,
+          currency: receipt.currencyAddress,
+          collateralApprovalTx: receipt.tx.transactionHash
+        });
+      }
+    });
```
```suggestion
    await Promise.map(Object.keys(this.contractCache), async contractAddress => {
      // The Perpetual requires approval to transfer the contract's collateral currency in order to post a bond.
      // We'll set this once to the max value and top up whenever the bot's allowance drops below MAX_INT / 2.
     const receipt = await setAllowance(
        this.web3,
        this.gasEstimator,
        this.account,
        contractAddress,
        this.contractCache[contractAddress].collateralAddress
      );
      if (receipt) {
        this.logger.info({
          at: "PerpetualProposer",
          message: "Approved Perpetual contract to transfer unlimited collateral tokens 💰",
          perpetual: receipt.spenderAddress,
          currency: receipt.currencyAddress,
          collateralApprovalTx: receipt.tx.transactionHash
        });
      }
    });
```

nicholaspai replied:
Done.

mrice32 commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,360 @@
+  }
+
+  async _cacheAndUpdatePriceFeeds() {
+    for (let contractAddress of Object.keys(this.contractCache)) {
```
nit: thoughts on doing this as a big `Promise.map` so we can run these inner async calls in parallel?

nicholaspai replied:
Nice catch definitely will do

mrice32 replied:
Bump.

nicholaspai replied:
Done

mrice32 reviewed (commented):
Flushing some comments. Will continue working through.

mrice32 commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,360 @@
+          this.commonPriceFeedConfig,
+          fundingRateIdentifier
+        );
+        this.priceFeedCache[fundingRateIdentifier] = priceFeed;
```
What if this comes back null (because we don't know how to construct this pf)? Should we send an error log? Do we just expect the bot to crash? It seems like we should probably just fire off an error log and keep going?

nicholaspai replied:
Yeah I think we should skip setting it in the cache during this method, and then emit an error log in the `updateFUndingRates()` call

mrice32 replied:
+1, that sounds like the right way to do it

mrice32 replied:
You don't even need to skip setting it here since it'll set it to null, and the other method should be able to recognize that null.

nicholaspai replied:
- I added an `if(pricefeed)` before the subsequent `await pricefeed.update()`
- I catch `!pricefeed` in the `updateFundingRates()` method and emit error log (and exit early) if no pricefeed.

mrice32 replied:
Bump.

mrice32 replied:
Ignore this^. LGTM

mrice32 commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,360 @@
+    // If the saved funding rate is not equal to the current funding rate within margin of error, then
+    // prepare request to update. We're assuming that the `offchainFundingRate` is the baseline or "expected"
+    // price.
+    let isPriceDisputable = isDeviationOutsideErrorMargin(
```
isPriceDisputable seems like the wrong name for this variable...? shoulduUdateFundingRate, maybe?

mrice32 replied:
Bump.

mrice32 commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,360 @@
+      return;
+    }
+
+    // If the saved funding rate is not equal to the current funding rate within margin of error, then
```
Probably don't need to do anything for this PR, but it could make sense for us to use a time/or ROI metric to determine how often to propose new funding rates. I think the current code makes sense for a bot that _doesn't care about profit_, but a profit-seeking bot would only care how many rewards they would get. This is aligned with UMA-run bots, but may not be as interesting for 3rd party bot runners (which may actually be a feature, not a bug since we'd like them to write their own bots!)

nicholaspai replied:
I'll add an issue!

chrismaree replied:
ye, agreed. could be tuned via a parameter that defines the "profit seeking" threshold of the bot. we'd just set this to zero but it can default for others to only do this in profitable contexts.

mrice32 reviewed (commented):
Looks awesome! A few more minor comments.

nicholaspai commented on packages/funding-rate-proposer/index.js:
```diff
@@ -0,0 +1,180 @@
+      // Common price feed configuration passed along to all those constructed by proposer.
+      commonPriceFeedConfig: process.env.COMMON_PRICE_FEED_CONFIG
+        ? JSON.parse(process.env.COMMON_PRICE_FEED_CONFIG)
+        : { lookback: 7200 },
```
Hm should we? Sholdn't it be inferred from the OptimisticOracle's proposal liveness? Regardless I'll add a TODO

nicholaspai replied:
Actually, I'm not sure there's any sort of lookback that makes sense, because the funding rate proposer is always checking out the current price of the funding rate, never historical. WDYT

mrice32 replied:
Yep, you're right. Nice catch. We probably don't want to set it to 0 or anything, so just leaving the default seems right.

mrice32 replied:
Bump.

nicholaspai replied:
I removed this

mrice32 reviewed (approved):
Looks great! I put Bump on all the threads that I think still need to be resolved. Otherwise, LGTM!

mrice32 commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,118 @@
+    this.lastBlockToSearchUntil = endingBlockNumber;
+    this.lastUpdateTimestamp = 0;
+
+    if (contractType != "ExpiringMultiPartyCreator" && contractType != "PerpetualCreator")
```
```suggestion
    if (contractType !== "ExpiringMultiPartyCreator" && contractType !== "PerpetualCreator")
```

nicholaspai replied:
Done

mrice32 reviewed (approved):
LGTM!

chrismaree commented on packages/disputer/test/index.js:
```diff
@@ -32,7 +32,7 @@ let pollingDelay = 0; // 0 polling delay creates a serverless bot that yields af
 let errorRetries = 1;
 let errorRetriesTimeout = 0.1; // 100 milliseconds between preforming retries
 let identifier = "TEST_IDENTIFIER";
-let fundingRateIdentifier = "TEST_FUNDiNG_IDENTIFIER";
```
GoOd CaTcH.

chrismaree commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,118 @@
+    startingBlockNumber = 0,
+    endingBlockNumber = null,
+    contractType = "PerpetualCreator"
+    // Default to PerpetualMultiParty for now since the first intended user is the funding rate proposer bot
```
```suggestion
    // Default to PerpetualMultiParty for now since the first intended user is the funding rate proposer bot.
```

nicholaspai replied:
Done

chrismaree commented on packages/financial-templates-lib/src/clients/FinancialContractFactoryEventClient.js:
```diff
@@ -0,0 +1,118 @@
+
+    if (contractType != "ExpiringMultiPartyCreator" && contractType != "PerpetualCreator")
+      throw new Error(
+        `Invalid contract type provided: ${contractType}! The financial product factory client only supports ExpiringMultiPartyCreator or PerpetualCreator`
```
```suggestion
        `Invalid contract type provided: ${contractType}! The financial product factory event client only supports ExpiringMultiPartyCreator or PerpetualCreator`
```

nicholaspai replied:
Actually ended up renaming this to just factory client, no event

chrismaree commented on packages/financial-templates-lib/src/helpers/allowances.js:
```diff
@@ -0,0 +1,26 @@
+const { getAbi } = require("@uma/core");
+
+// Sets `owner` allowance for `spender` to MAX, unless `spender` already has an allowance > 1/2 of the MAX_UINT.
+// Return array of successful approval transactions or undefined for skipped approvals.
```
this comment implies this can do multiple approvals in one go. Wont it only do one at a time?

nicholaspai replied:
Yup you're right, will fix comment

chrismaree commented on packages/financial-templates-lib/src/helpers/math.js:
```diff
@@ -0,0 +1,37 @@
+const assert = require("assert");
```
🤩 

Will be more to add to this in the future for sure. thanks for the initial refactor!

chrismaree commented on packages/financial-templates-lib/src/price-feed/DefaultPriceFeedConfigs.js:
```diff
@@ -557,30 +557,6 @@ const defaultConfigs = {
     minuteLookback: 7200,
     hourlyLookback: 604800,
     minTimeBetweenUpdates: 60
-  },
```
why was uSTONKS removed?

nicholaspai replied:
Oops I'll merge this back in! Great catch

chrismaree commented on packages/financial-templates-lib/test/truffle-fixture.js:
```diff
@@ -0,0 +1,2 @@
+// Export the truffle-fixture file from `/core/test` so that we can use the `.deployed()` syntax in our tests.
```
where are you using this? are you using this in other packages? IMHO this is something we should strictly _not_ do. This somewhat defeats a lot of the gains we get from using hardhat. You dont want to re-run all the migrations for a test and rather deploy what you need in the tests. I think this should be the pattern that we follow for all packages that are _not_ `core`. This is how other packages are set up right now. Liquidator/disputer/monitor dont rely on any truffle migrations for this exact reason.

chrismaree replied:
Looking further down, I see you are doing things like
```
timer = await Timer.deployed();
perpFactory = await PerpetualCreator.deployed();
```

IMHO you should refactor this to _not_ be dependent on the core migrations. this will drastically speed up your unit tests and follow the patterns in the other non-core packages.

nicholaspai replied:
@chrismaree I need this because the `EMP/PerpCreator` contracts can't be deployed unless I link their their EMPLib/PerpLib contracts which I can't seem to figure out how to do in hardhat

nicholaspai replied:
Ended up fixing this by adding `await PerpetualCreator.link(await PerpetualLib.new())`

chrismaree commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,370 @@
+        });
+      } catch (error) {
+        const message =
+          error.type === "call"
```
nice

chrismaree commented on packages/funding-rate-proposer/src/proposer.js:
```diff
@@ -0,0 +1,370 @@
+    }
+  }
+  // Sets allowances for all collateral currencies used live perpetual contracts.
+  async _setAllowances() {
```
I like this

chrismaree reviewed (approved):
This is a beast of a PR! excellent work on getting it out so quickly and with such high quality, well commented code!

LGTM, granting you make a refactor I requested on the migration-reuse from core.

nicholaspai commented:
@chrismaree @mrice32 I've addressed all of your comments PTAL when you have time! Any leftover issues are collected in this [project board card](https://github.com/UMAprotocol/protocol/projects/4#card-46767632)

mrice32 reviewed (approved):
LGTM!
